Fix: avoid long thread sleeps on codec close

kares opened the pull request:
The PR fixes a stalled shutdown process due `codec.close`, details from https://github.com/logstash-plugins/logstash-input-file/issues/298:
> In rare cases (multiple file inputs helps reproduce the issue) as the plugin attempts to stop it's identity codec, it ends up waking up the cleaner thread but the thread gets right back to (a [5 minute](https://github.com/logstash-plugins/logstash-codec-multiline/blob/v3.0.10/lib/logstash/codecs/identity_map_codec.rb#L104)) sleep (when the [`cleaner.stop`](https://github.com/logstash-plugins/logstash-codec-multiline/blob/v3.0.10/lib/logstash/codecs/identity_map_codec.rb#L70) -> [`wakeup`](https://github.com/logstash-plugins/logstash-codec-multiline/blob/v3.0.10/lib/logstash/codecs/identity_map_codec.rb#L70) happens right before [`sleep`](https://github.com/logstash-plugins/logstash-codec-multiline/blob/v3.0.10/lib/logstash/codecs/identity_map_codec.rb#L53) on the other thread). 

There's a more technical (isolated) reproducer which served as the inspiration for the spec: https://github.com/logstash-plugins/logstash-codec-multiline/issues/68

There's also a minor refactoring to explicitly declare the concurrent dependency + avoid the legacy `ThreadSafe::Hash` usage in favor of its `Concurrent::Hash` counterpart.

closes https://github.com/logstash-plugins/logstash-codec-multiline/issues/68

kares committed:
Refactor: drop (legacy) ThreadSafe::Hash usage

kares committed:
Fix: avoid long sleeps on codec close

kares committed:
Chore: cleanup comment

jsvd commented on lib/logstash/codecs/identity_map_codec.rb:
```diff
@@ -48,12 +48,16 @@ def initialize(listener, interval, method_symbol)
       @running.make_true
-      @thread = Thread.new() do
+      @thread = Thread.start do
+        Thread.current.tap do |thread|
+          # to be able to distinguish cleaner from auto_flusher in thread dumps
+          thread.name = @method_symbol.to_s if thread.respond_to?(:name)
+        end
```
We can leverage [`LogStash::Util.set_thread_name()`](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/util.rb#L27) here:

```suggestion
        # to be able to distinguish cleaner from auto_flusher in thread dumps
        LogStash::Util.set_thread_name(@method_symbol.to_s)
```

kares replied:
Thanks, that wasn't the point though ... the Java name set by the JRuby is fine, since it includes the .rb path + :line. 
Setting the name at the Ruby level just adds extra information to the Java thread naming format used by the runtime.

kares replied:
changed to use the LS api instead setting the Ruby name, here's how the change looks as the Java level:
previous (`thread.name = ...`):
`"Ruby-0-Thread-1@map_cleanup: .../logstash-codec-multiline/lib/logstash/codecs/identity_map_codec.rb:51"`

new thread name:
`"IdentityMapCodec#map_cleanup"`

let me know if you have ideas for a better name a sole 'cleaner' just seemed more confusing ...

kares committed:
Refactor: set Java thread name directly

previous format:
"Ruby-0-Thread-1@map_cleanup:
.../logstash-codec-multiline/lib/logstash/codecs/identity_map_codec.rb:51"

new format:
"IdentityMapCodec#map_cleanup"

jsvd reviewed (approved):
LGTM

mback2k commented:
Hi there, since I am the customer initially reporting this issue and thankful for a solution, would you be so kind and tell me to which version I have to update to get this fix? Thanks in advance!

kares commented:
Hey Marc, sorry to cause confusion. 
I forgot to include the version + changelog entry in the PR but it's linked back from the [following commit](https://github.com/kares/logstash-codec-multiline/commit/efaa201f9893838c8c75ee05cc24ce3700736033).
Just do a `bin/logstash-plugin update logstash-codec-multiline` (should than update the plugin to >= 3.0.11)
